Issue/10784 blaze analytics part ii

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #10784 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Ads tracking for all the new events for Blaze campaign creation native flow

### Testing instructions
<!-- Step-by-step testing instructions. When necessary, break out individual scenarios that need testing, and consider including a checklist for the reviewer to go through. -->
1. Trigger Blaze creation flow 
2. Follow the table from the tracking section for Blaze i3 PT: pe5sF9-2mb-p2 and check in the Android logcat for the following output: `🔵 Tracked: event_name, Properties: property_name:value ...` to ensure all of the events from the table are fired appropriately.

Note: the last 2 events from the table `_blaze_campaign_creation_success` and `_blaze_campaign_creation_failed` are not included in this PR as the code for firing the campaign creation and handling the success/error result is not ready yet

JorgeMucientes committed:
Track when Learn mor is clicked from Blaze intro screen

JorgeMucientes committed:
Track campaign preview screen displayed

JorgeMucientes committed:
Track events for campaign preview screen

JorgeMucientes committed:
Track events for edit Ad screen

JorgeMucientes committed:
Track events for budget screen

JorgeMucientes committed:
Track events for saved target options

JorgeMucientes committed:
Track events for saving campaign targeting options

JorgeMucientes committed:
Track events for payment screens

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes committed:
Merge branch 'issue/10784-blaze-analytics-part-I' into issue/10784-blaze-analytics-part-II

JorgeMucientes committed:
Fix detekt: remove unused import

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -128,6 +140,12 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
             )
         }
         fetchAdForecast()
+        analyticsTrackerWrapper.track(
```
This gets called every time the duration slider is moved, which seems like a bug to me, I started a discussion about it here p1708002828270189-slack-C03L1NF1EA3

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/destination/BlazeCampaignCreationAdDestinationViewModel.kt:
```diff
@@ -64,6 +67,7 @@ class BlazeCampaignCreationAdDestinationViewModel @Inject constructor(
     }
 
     fun onDestinationUrlChanged(destinationUrl: String) {
+        analyticsTrackerWrapper.track(stat = BLAZE_CREATION_EDIT_DESTINATION_SAVE_TAPPED)
```
There is a small difference between iOS and Android here, and I think we might need to have different events because of it, in iOS the `Save` button is for the whole screen, and saves both the targetUrl and the parameters, while here this gets called when the targetUrl is changed, and not the parameters, WDYT?

JorgeMucientes replied:
Good point, @hichamboushaba. I actually had thoughts of initiating a discussion precisely about this here p1707952047198469-slack-C03L1NF1EA3 😅. However, I refrained from it because I didn't think the difference was that relevant. But after reading your comment I guess I should reconsider it :) 

I agree we should have different events here as the frequency in which the save button is clicked will likely be very different between platforms. 

Wdyt about a new event for Android: `blaze_creation_destination_url_changed` replacing the actual `blaze_creation_edit_destination_save_tapped`, and leaving the latter only on iOS? 

If agreed, I can proceed and update the P2 with this slight difference between platforms

hichamboushaba replied:
> Wdyt about a new event for Android: blaze_creation_destination_url_changed replacing the actual blaze_creation_edit_destination_save_tapped, and leaving the latter only on iOS?

Sounds good, will you add one other event for the URL parameters?

JorgeMucientes replied:
I don't mind. Is a really minor change, but do you think we need that level of detail? Are we going to use that event? 

Nah, let's add it just in case; it won't harm. Although I have doubts about how useful it will be.

hichamboushaba replied:
IMO this is the part that most users will use, the majority will be happy about the product URL as destination, so they probably won't change.

On another note, just an idea, if we want to keep the same event as iOS, we can probably just fire it when the user taps back on the screen, we can check if the data has changed, and if yes, we consider it as a save button call, it's a minor point, just wanted to share it, feel free to ignore it though if you already added the new events.

JorgeMucientes replied:
Yeah good point, I like that idea better, especially because it keeps event parity between platforms. Applied it here: https://github.com/woocommerce/woocommerce-android/pull/10820/commits/5a9474c2442bb03d533c9f03c03a0cefc88fd5cb

hichamboushaba reviewed (approved):
Nice work @JorgeMucientes, all of the events work well, I left two comments, but none is blocking, we can address them if needed on separate PRs.

JorgeMucientes committed:
Merge branch 'trunk' into issue/10784-blaze-analytics-part-II

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/ad/BlazeCampaignCreationEditAdViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/payment/BlazeCampaignPaymentSummaryViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/preview/BlazeCampaignCreationPreviewViewModel.kt

JorgeMucientes committed:
Update only campaign duration once the apply button is clicked

JorgeMucientes committed:
Track duration applied event only after clicking on apply button from bottom sheet

JorgeMucientes committed:
Fix bug, ensure total budget value is within min and max ranges

JorgeMucientes committed:
Merge branch 'trunk' into issue/10784-blaze-analytics-part-II

JorgeMucientes committed:
Track edit destination event tapped event only if destination changes

JorgeMucientes committed:
Track edit destination event tapped event only if destination changes

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -219,11 +223,13 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
         val dailySpending: String,
         val forecast: ForecastUi,
         val durationInDays: Int,
+        val sliderDurationValue: Float,
```
I know @JorgeMucientes you are still working on this, just wondering, do we need an additional property in the State for this, when I was planning to work on this I was think we can just make this a UI detail, and handle it using Compose's functions, so we'll just initialize a State using `rememberSaveable` in the BottomSheet to keep the state of the duration slider, then when `apply` is clicked, we pass this State's value as a callback to our ViewModel, WDYT?

JorgeMucientes committed:
Use compose property to keep track of the duration slider value

JorgeMucientes committed:
Track if AI content was used for the creating the ad

In addition this change adds a repository cache for ai ad suggestions to avoid requesting suggestions twice when entering edit ad screen

JorgeMucientes commented:
Hey @hichamboushaba, the PR is ready for another round. Thanks for all the provided feedback 👍🏼. Here's a summary of the changes applied: 
- Update only campaign duration once the `Apply` button is clicked: https://github.com/woocommerce/woocommerce-android/pull/10820/commits/177c8272a9593c27eeb93765634ba74340a14b7a and then followed the suggestion to simplify the code: https://github.com/woocommerce/woocommerce-android/pull/10820/commits/7deed09411205f7c84cacf4832b38c9d365f7675
- Track `blaze_creation_edit_destination_save_tapped` when navigating back from the destination screen and the url or any of its params has changed: https://github.com/woocommerce/woocommerce-android/pull/10820/commits/5a9474c2442bb03d533c9f03c03a0cefc88fd5cb
- Fixes bug. When changing the campaign duration, the total budget value was not updated to fit in between the allowed minimum and maximum daily spend: https://github.com/woocommerce/woocommerce-android/pull/10820/commits/b1f3293d7436988499dab96cc892812a6802ecdd
- Added missing property `is_ai_suggested_ad_content:true/false` to the event `_blaze_creation_confirm_details_tapped`. While adding this, I realized we were fetching AI suggestions every time we entered the preview screen and again when opening the edit ad screen. I've added an in-memory cache to avoid this and to have ai suggestions accessible to set the correct value for the `is_ai_suggested_ad_content` property:  https://github.com/woocommerce/woocommerce-android/pull/10820/commits/20bf87d7cf7efc98dd53e0944cbd6bde96a080b3

Sorry for the many changes added, but I think applying them in this PR made sense because that enabled firing the tracking events at the right time.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt:
```diff
@@ -23,9 +23,11 @@ import org.wordpress.android.fluxc.model.blaze.BlazeTargetingParameters
 import kotlin.math.roundToInt
 import kotlin.time.Duration.Companion.days
 
+@Singleton
```
np, Do we need to make this `@Singleton` for this? this will cause the data to be cached for the whole lifetime of the app, and if someone forgets to pass `forceRefresh`, we could be reusing wrong suggestions.

WDYT about saving the suggestions in the Preview ViewModel, and pass them as an argument to Edit Ad ViewModel?

JorgeMucientes replied:
Fair enough, the reason I went for this approach initially is because it was easier and required a bit less file changes. But its true that it has that risk. So I used nav args instead and completely removed the repository cache.
Good suggestion 👍🏼

hichamboushaba replied:
Thanks @JorgeMucientes, and sorry for nitpicking on this.
One point just for context, no need to change unless you agree with me, I suggested caching the suggestion at the ViewModel level, and not the `CampaignDetails`, because adding them as a parameter to the `CampaignDetails` will cause them to be saved to the SavedState, while we don't need it really, as they can be re-fetched if the screen ViewModel gets re-created.

JorgeMucientes replied:
Makes sense. I've extracted it out of the `CampaignDetails` model to a class field 👍🏼

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -167,10 +166,20 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
+        }
+    }
+
+    private fun calculateDailySpending(duration: Int): Float {
```
Nice find, thanks for fixing this.

JorgeMucientes replied:
Fair enough, the reason I went for this approach initially is because it was easier and required a bit less file changes. But its true that it has that risk. So I used nav args instead and completely removed the repository cache. 
Good suggestion 👍🏼

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -167,10 +166,20 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
+
+    private fun calculateDailySpending(duration: Int): Float {
+        val dailySpend = budgetUiState.value.totalBudget / duration
+        return if (dailySpend < CAMPAIGN_MINIMUM_DAILY_SPEND) {
+            CAMPAIGN_MINIMUM_DAILY_SPEND
+        } else if (dailySpend > CAMPAIGN_MAXIMUM_DAILY_SPEND) {
+            CAMPAIGN_MAXIMUM_DAILY_SPEND
+        } else {
+            dailySpend
         }
```
np, we can simplify this logic using [`coerceIn`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/coerce-in.html).

JorgeMucientes replied:
Nice! Way simpler TIL

hichamboushaba reviewed (approved):
Nice work @JorgeMucientes 👏, thanks for addressing those issues and the new ones.

I left some np comments, but please check the first one before merging.

JorgeMucientes committed:
Simplify code by using ext function coerceIn

JorgeMucientes committed:
Remove ai ad suggestions cached and pass them as nav args

JorgeMucientes committed:
Merge branch 'trunk' into issue/10784-blaze-analytics-part-II

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/destination/BlazeCampaignCreationAdDestinationViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/payment/BlazeCampaignPaymentSummaryViewModel.kt

JorgeMucientes committed:
Fix detekt issues after merge with trunk

JorgeMucientes committed:
Fix unit tests

JorgeMucientes committed:
Merge branch 'trunk' into issue/10784-blaze-analytics-part-II

JorgeMucientes committed:
Remove ai suggestions from Campaign details model¥

JorgeMucientes committed:
Fix redundant qualifiers and wrong imports
